### DISCUSSION TITLE
How to use the "UniformBuffer" struct?

### DISCUSSION THREAD

**Discussion by lo-vi** [bevyengine/bevy#16714] [Q&A] [answered]
Hello! I've discovered Bevy recently and I'm trying to write a compute shader using the "GLSL compute shaders" guide. I've semi-successfully rendered the colored image to the screen but I can't for the life of me figure out how to use the [UniformBuffer](https://docs.rs/bevy/latest/bevy/render/render_resource/struct.UniformBuffer.html) struct.

I've created the uniform buffer.
````
#[derive(Resource, Clone, ExtractResource, Default, ShaderType)]
pub struct TimeBuffer {
    pub time: f32,
    #[cfg(feature = "webgl2")]
    _webgl2_padding: Vec3
}

impl TimeBuffer {
     pub fn create_uniform(&self) -> UniformBuffer<&Self> {
        UniformBuffer::from(self)
    }

    pub fn setup(mut commands: Commands) {
        commands.insert_resource(Self::default());
    }
}
````

The problem occurs when I try to bind the uniform buffer to the shader and get a `Failed to get buffer` error.
````
fn prepare_bind_group(
    mut commands: Commands,
    pipeline: Res<ComputePipeline>,
    render_device: Res<RenderDevice>,
    image: Res<DisplayImage>,
    images: Res<RenderAssets<GpuImage>>,
    buffer: Res<TimeBuffer>
) {
    let image = images.get(&image.0).unwrap();
    let uniform = buffer.create_uniform();

    let bind_group = render_device.create_bind_group(
        None,
        &pipeline.layout,
        &BindGroupEntries::with_indices((
            (0, image.texture_view.into_binding()),
            (1, uniform.into_binding())
        ))
    );

    commands.insert_resource(Self(bind_group))
}
````
````
let layout = render_device.create_bind_group_layout(
    None,
    &BindGroupLayoutEntries::with_indices(
        ShaderStages::COMPUTE,
            (
                (0, texture_storage_2d(TextureFormat::Rgba8Unorm, StorageTextureAccess::WriteOnly)),
                (1, uniform_buffer::<TimeBuffer>(false))
            )
    )   
);
````

I've can't seem to find any examples using just [UniformBuffer](https://docs.rs/bevy/latest/bevy/render/render_resource/struct.UniformBuffer.html). The "Post Processing" example uses [DynamicUniformBuffer](https://docs.rs/bevy/latest/bevy/render/render_resource/struct.DynamicUniformBuffer.html) which binds using...
````
let settings_uniforms = world.resource::<ComponentUniforms<PostProcessSettings>>();
let Some(settings_binding) = settings_uniforms.uniforms().binding() else {
    return Ok(());
};

````
inside the `PostProcessingNode.run` method but the size of my data in the buffer won't be changing. I would use the [GlobalUniforms](https://docs.rs/bevy/latest/bevy/render/globals/struct.GlobalsUniform.html) import but only use case for that is only related to time.

Is there something I'm missing here?  Any examples/help on creating and updating [UniformBuffer](https://docs.rs/bevy/latest/bevy/render/render_resource/struct.UniformBuffer.html) would be greatly appreciated!

**Comment by JMS55** [comment, marked as answer]
You need to call https://docs.rs/bevy/latest/bevy/render/render_resource/struct.UniformBuffer.html#method.write_buffer to schedule a flush of the data from CPU->GPU.

**Comment by Touma-Kazusa2** [comment]
The "Post Processing" example uses UniformComponentPlugin,it registers prepare_uniform_components system in RenderSet::PrepareResources,and it uses the write() function to write buffer
```
impl<'a, T: ShaderType + WriteInto> DynamicUniformBufferWriter<'a, T> {
    pub fn write(&mut self, value: &T) -> u32 {
        self.buffer.write(value).unwrap() as u32
    }
}
```

**Comment by lo-vi** [comment]
Slight oversight from me, thank you guys so much!
